moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java

```diff
@@ -0,0 +1,218 @@
+        var bindings = new LinkedHashMap<String, ComponentBinding<?>>();
+        var constructorParams = new Class<?>[components.length];
+        var lookup = MethodHandles.lookup();
+        for (int i = 0, componentsLength = components.length; i < componentsLength; i++) {
+          RecordComponent component = components[i];
+          constructorParams[i] = component.getType();
+          var name = component.getName();
+          var componentType = component.getGenericType();
+          if (componentType instanceof TypeVariable<?> typeVariable) {
+            var typeVarName = typeVariable.getName();
+            if (mappedTypeArgs == null) {
+              throw new AssertionError(
+                  "No mapped type arguments found for type '" + typeVarName + "'");
+            }
+            var mappedType = mappedTypeArgs.get(typeVarName);
+            if (mappedType == null) {
+              throw new AssertionError(
+                  "No materialized type argument found for type '" + typeVarName + "'");
+            }
+            componentType = mappedType;
+          }
+          var jsonName = name;
+          Set<Annotation> qualifiers = null;
+          for (var annotation : component.getDeclaredAnnotations()) {
+            if (annotation instanceof Json jsonAnnotation) {
+              jsonName = jsonAnnotation.name();
+            } else {
+              if (annotation.annotationType().isAnnotationPresent(JsonQualifier.class)) {
+                if (qualifiers == null) {
+                  qualifiers = new LinkedHashSet<>();
+                }
+                qualifiers.add(annotation);
+              }
+            }
+          }
+          if (qualifiers == null) {
+            qualifiers = Collections.emptySet();
+          }
+          var adapter = moshi.adapter(componentType, qualifiers);
+          MethodHandle accessor;
+          try {
+            accessor = lookup.unreflect(component.getAccessor());
+          } catch (IllegalAccessException e) {
+            throw new AssertionError(e);
+          }
+          var componentBinding = new ComponentBinding<>(name, jsonName, adapter, accessor);
+          var replaced = bindings.put(jsonName, componentBinding);
+          if (replaced != null) {
+            throw new IllegalArgumentException(
+                "Conflicting components:\n"
+                    + "    "
+                    + replaced.name
+                    + "\n"
+                    + "    "
+                    + componentBinding.name);
+          }
+        }
+
+        MethodHandle constructor;
+        try {
+          constructor = lookup.findConstructor(rawType, methodType(void.class, constructorParams));
+        } catch (NoSuchMethodException | IllegalAccessException e) {
+          throw new AssertionError(e);
+        }
+        return new RecordJsonAdapter<>(constructor, rawType.getSimpleName(), bindings).nullSafe();
+      };
+
+  private static record ComponentBinding<T>(
```

Comment:
yo dog

Reply:
yeah I think the static is implicit but the IDE didn't give me a warning about it being redundant